File: algorithms/117. Populating Next Right Pointers in Each Node II/solution.cc
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        TreeLinkNode dummy(0);
        auto tail = & dummy;
        while ( root ) {
            tail->next = root->left;
            if ( tail->next ) tail = tail->next;
            tail->next = root->right;
            if ( tail->next ) tail = tail->next;
            root = root->next;
            if ( ! root ) {
                root = dummy.next;
                tail = & dummy;
            }
        }
    }
};

File: algorithms/117. Populating Next Right Pointers in Each Node II/solution.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        tail = dummy = TreeLinkNode(0)
        while root:
            tail.next = root.left
            tail = tail.next or tail
            tail.next = root.right
            tail = tail.next or tail
            root = root.next
            if not root:
                root = dummy.next
                tail = dummy
